uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js

```diff
@@ -0,0 +1,591 @@
+                $scope.$digest();
+              });
+
+              it('will update change amount', function () {
+                expect($ctrl.balance.change.amount).toEqual(-2);
+              });
+
+              it('will update closing balance', function () {
+                expect($ctrl.balance.closing).toEqual(-2);
+              });
+            });
+          });
+        });
+
+        describe('change details', function () {
+
+          describe('hidden', function () {
+
+            it('by default', function () {
+              expect($ctrl.uiOptions.isChangeExpanded).toBeFalsy();
+            });
+          });
+
+          describe('when expanded', function () {
+
+            beforeEach(function () {
+              $ctrl.uiOptions.isChangeExpanded = true;
+            });
+
+            it('will display balance breakdown', function () {
+              expect($ctrl.uiOptions.isChangeExpanded).toBeTruthy();
+            });
+
+            describe('during pagination', function () {
+
+              describe('init', function () {
+
+                it('should', function () {
+                  expect($ctrl.pagination.totalItems).toEqual(0);
+                });
+              });
+
+              describe('after date is selected', function () {
+
+                beforeEach(function () {
+                  $ctrl.uiOptions.toDate = $ctrl.uiOptions.fromDate = new Date();
+                  $ctrl.onDateChange($ctrl.uiOptions.fromDate, true);
+                  $ctrl.onDateChange($ctrl.uiOptions.toDate, false);
+                  $scope.$digest();
+                });
+
+                it('will select default page', function () {
+                  expect($ctrl.pagination.currentPage).toEqual(1);
+                });
+
+                it('should set totalItems', function () {
+                  expect($ctrl.pagination.totalItems).toBeGreaterThan(0);
+                });
+
+                describe('change selected page', function () {
+                  var beforeFilteredItems;
+
+                  beforeEach(function () {
+                    beforeFilteredItems = $ctrl.pagination.filteredbreakdown;
+                    $ctrl.pagination.currentPage = 2;
+                    $ctrl.pagination.pageChanged();
+                  });
+
+                  it('should change current page', function () {
+                    expect($ctrl.pagination.currentPage).not.toEqual(1);
+                  });
+
+                  it('should change filtered data', function () {
+                    expect($ctrl.pagination.filteredbreakdown[0]).not.toEqual(beforeFilteredItems[0]);
+                  });
+                });
+              });
+            });
+          });
+        });
+      });
+
+      describe('save leave request', function () {
+
+        describe('with invalid fields', function () {
+
+          it('does not have required fields', function () {
+            expect($ctrl.leaveRequest.from_date).not.toBeDefined();
```

Comment:
I don't understand this test, you are not even submitting the leave request? The title seems incoherent ("save leave request / with invalid fields / does not have required fields") and in the end this looks simply an initialization test, since again you are not saving the leave request before testing the expectation